### DISCUSSION TITLE
Use HEAD requests to avoid being rate limited by DockerHub

### DISCUSSION THREAD

**Discussion by tripplet** [containrrr/watchtower#631] [Feature Requests] [answered]
Based on how exactly manifest requests will be counted, how many containers are monitored and which interval is configured this will probably affect many watchtower users.
[https://www.docker.com/blog/scaling-docker-to-serve-millions-more-developers-network-egress/](https://www.docker.com/blog/scaling-docker-to-serve-millions-more-developers-network-egress/)
Hacker News [discussion](https://news.ycombinator.com/item?id=24262757)

**Comment by zoispag** [comment]
Maybe we need to rethink about the skopeo solution @simskij @piksel

**Reply by simonaronsson** [reply]
I totally agree, we also need to consider moving watchtower itself to GitHub packages.

**Reply by victorcmoura** [reply]
Yeah, now it sounds more interesting =S

**Reply by prabu-ssb** [reply]
Yes, skopeo (#607) might be a good long term solution.

**Reply by piksel** [reply]
I should just add  that we pretty much ruled out skopeo since it doesn't keep track of the digests for the manifests and will do a full rate-limited manifest GET for every query, so it doesn't help us much.

**Comment by simonaronsson** [comment, marked as answer]
I've tried to answer this in more detail in https://github.com/containrrr/watchtower/discussions/668#discussioncomment-116021, but the executive summary is that we don't need to change anything, the docker client will take care of this for us and make sure requests counting towards the rate limit only happens when the image digest has changed. 🎉

**Reply by piksel** [reply]
Turns out this wasn't the full story. Docker sometimes doesn't do this, but performs a full GET request when checking for changes.
The most up to date discussion is currently in #669.

**Comment by willheineman** [comment]
Since this went live today, I am seeing the rate limit errors appearing frequently. I have 16 containers that I run on a home server, and most of the images I run don't even update every 24 hours. But starting at about 12pm EST I am getting the following error:

`toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit. Proceeding to next.`

I am going to try to change the frequency to 60 minutes, so that I can stay under the rate limit (16 images * 6 pulls per period = 96 every 6 hours)

But is there anything else I should be doing? I am looking into making authenticated calls, which would let me go to every 30 minutes.

**Reply by simonaronsson** [reply]
Not at this moment, I'm afraid. We're currently tracking this issue in #669 if you'd like to stay informed.

**Comment by nhgiang196** [comment]
Despite the pull rate limit message. Watchtower in my computer mangage to update the exactly image by polling every 30 seconds.
![p3DP0VAovx](https://user-images.githubusercontent.com/22267870/98620136-43a51d80-2337-11eb-857d-948dd6dd4838.png)

The only problem is I got TOO MANY emails that said:...
![37VvvMa](https://user-images.githubusercontent.com/22267870/98619875-b1047e80-2336-11eb-9d7c-1a04694ebcbf.png)

**Comment by simonaronsson** [comment]
We've moved discussions on this topic to #669. Locking this discussion to make sure everything is available at the same place.
